pkg/reduce/readloop/ordered.go

```diff
@@ -0,0 +1,169 @@
+package readloop
+
+import (
+	"container/list"
+	"context"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/numaproj/numaflow/pkg/isb"
+	"github.com/numaproj/numaflow/pkg/isb/forward"
+	"github.com/numaproj/numaflow/pkg/watermark/publish"
+
+	"github.com/numaproj/numaflow/pkg/pbq"
+	"github.com/numaproj/numaflow/pkg/pbq/partition"
+	"github.com/numaproj/numaflow/pkg/reduce/pnf"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	udfreducer "github.com/numaproj/numaflow/pkg/udf/reducer"
+)
+
+var retryDelay = 1 * time.Second
+
+// task wraps the `ProcessAndForward`.
+type task struct {
+	// doneCh is used to notify when the task has been completed.
+	doneCh chan struct{}
+	pf     *pnf.ProcessAndForward
+}
+
+// orderedForwarder orders the forwarding of the result of the execution of the tasks, even though the tasks itself are
+// run concurrently in an out of ordered fashion.
+type orderedForwarder struct {
```

Comment:
Can you give more info about why it needs to be ordered?

Reply:
After a reduce function, the result's watermark is based on the end-time of the window. Since multiple keyed-windows are being processed concurrently, and a newer keyed-window gets completed before an older one, the assignment of the watermark will be wrong.

The closing of keyed-windows (partitions) are in order based on watermark progression, but the result need not be because one partition can be slower than the other based on the user's reduce logic.

Reply:
👍